### PR TITLE
[AAP-74719] feat(ci): add SonarCloud analysis workflow

### PR DESCRIPTION
**Pull request by tznamena**
## Description

- **What is being changed?** Adds a SonarCloud analysis GitHub Actions workflow (`sonar-pr.yml`) that performs static code analysis and coverage reporting via SonarCloud.
- **Why is this change needed?** To enable automated code quality scanning on PRs and long-lived branches, as tracked in AAP-74719.
- **How does this change address the issue?** The workflow is triggered after the Unit Tests workflow finishes. A gate job checks two repository variables (`SONAR_BRANCHES` and `SONAR_TOKEN_SECRET_NAME`) before allowing analysis to proceed. This design makes the same workflow file work across multiple repositories (upstream and downstream) without hardcoding repository names, and safely skips execution in forks that lack the required configuration.

## Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Configuration change

## Self-Review Checklist

- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [x] I have updated documentation where needed
- [x] I have considered the security impact of these changes
- [x] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [ ] I have tested the changes in my local environment

### Prerequisites
- Repository admin access to configure repository variables

### Steps to Test
1. Set repository variable `SONAR_BRANCHES` to `devel,feature-*` (or desired branch patterns)
2. Set repository variable `SONAR_TOKEN_SECRET_NAME` to the name of the secret holding the SonarCloud token
3. Open a PR targeting a matching branch — the SonarCloud PR Analysis job should run after Unit Tests completes
4. Push to a matching branch — the SonarCloud Branch Analysis job should run after Unit Tests completes
5. Verify that forks without the variables configured see all jobs skipped (not failed)

### Expected Results
- PR analysis runs and reports findings on the PR in SonarCloud
- Branch analysis runs on push to configured branches
- Forks and unconfigured repos skip cleanly with no failures

### Required Actions
- [ ] Requires infrastructure/deployment changes
  
- [ ] Requires downstream repository changes
  

## Summary by CodeRabbit

* **Chores**
  * Automated code quality scanning is now integrated into the CI/CD pipeline, running after unit tests complete for pull requests and branch pushes with aggregated coverage analysis.

### CODE REVIEW

File: .github/workflows/sonar-pr.yml [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        run: |
+          if [ "$EVENT_TYPE" = "pull_request" ]; then
+            # For PRs we need the base branch, but workflow_run doesn't expose it directly.
+            # We allow all PRs through the gate; the base branch is validated implicitly
+            # because the PR targets a branch that itself must be in SONAR_BRANCHES to
+            # receive branch analysis. PR analysis focuses on changed code quality.
+            echo "branch=$HEAD_BRANCH" >> "$GITHUB_OUTPUT"
+            echo "allowed=true" >> "$GITHUB_OUTPUT"
+            echo "✅ PR analysis is always allowed when SONAR_BRANCHES is configured"
+            exit 0
```
[the remark this answers is not in the document]
Comment (tznamena) [reply .github/workflows/sonar-pr.yml:88]: Addressed.

File: .github/workflows/sonar-pr.yml [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+      contents: read
+      actions: read
+      pull-requests: read
+    if: |
+      needs.gate.outputs.allowed == 'true' &&
+      github.event.workflow_run.event == 'pull_request'
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
```
[the remark this answers is not in the document]
Comment (tznamena) [reply .github/workflows/sonar-pr.yml:126]: Addressed.

File: .github/workflows/sonar-pr.yml [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+          echo "PR_BASE=$PR_BASE" >> $GITHUB_ENV
+          echo "PR_HEAD=$PR_HEAD" >> $GITHUB_ENV
+          echo "COMMIT_SHA=$COMMIT_SHA" >> $GITHUB_ENV
+          echo "REPO_NAME=$REPO_NAME" >> $GITHUB_ENV
+
+          # Get all changed files from PR (with error handling)
+          files=""
+          if [ -n "$PR_NUMBER" ]; then
+            if gh api repos/$REPO_NAME/pulls/$PR_NUMBER/files --jq '.[].filename' > /tmp/pr_files.txt 2>/tmp/pr_error.txt; then
+              files=$(cat /tmp/pr_files.txt)
```
[the remark this answers is not in the document]
Comment (tznamena) [reply .github/workflows/sonar-pr.yml]: Addressed.

**Review by john-westcott-iv — APPROVED** [review]
## Code Review: Add SonarCloud analysis workflow

**Verdict**: LGTM

Nice improvement over the aap-gateway/stable-2.5 reference — the variable-based gate (`SONAR_BRANCHES` / `SONAR_TOKEN_SECRET_NAME`) is much better than hardcoded repository names, `fetch-depth: 0` is correct for SonarCloud blame analysis, and fork safety is cleaner.

Two nits:

**1. Add `--paginate` to the PR files API call (line 193).** Without it, `SONAR_INCLUSIONS` is truncated for PRs with 30+ changed files. Same issue exists in the aap-gateway reference. One-word fix:
```diff
- if gh api repos/$REPO_NAME/pulls/$PR_NUMBER/files --jq '.[].filename' ...
+ if gh api --paginate repos/$REPO_NAME/pulls/$PR_NUMBER/files --jq '.[].filename' ...
```

**2. Consider pinning checkout to `workflow_run.head_sha`.** Both checkout steps default to the repo's default branch in `workflow_run` context. The `gh pr checkout` step mitigates this for PR analysis but could pick up newer commits than what Unit Tests ran against. The aap-gateway reference has the same issue, so this isn't new — but adding `ref: ${{ github.event.workflow_run.head_sha }}` would be the correct fix for both.

**Comment by tznamena** [comment]
@john-westcott-iv Those items have been addressed, I guess you were reviewing as I was working on fixes.
